### PR TITLE
`selector-attribute-square-brackets-space-inside` rule.

### PR DESCRIPTION
**Pull request by evilebottnawi**

### CODE REVIEW

File: src/rules/selector-attribute-square-brackets-space-inside/index.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      ],
+    })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      selectorParser(selectorTree => {
+        selectorTree.eachAttribute(attributeNode => {
+          const attributeSelectorString = attributeNode.toString()
+
+          styleSearch({ source: attributeSelectorString, target: "[" }, match => {
```
Comment (davidtheclark) [inline src/rules/selector-attribute-square-brackets-space-inside/index.js]: As over here: https://github.com/stylelint/stylelint/pull/1079#discussion_r60407832 --- I'd bet the spacing data is already in the node, so the two searches might be redundant.

Comment (evilebottnawi) [inline src/rules/selector-attribute-square-brackets-space-inside/index.js]: @davidtheclark 

```css
[ target=blank ] { }
```

```shell
Attribute {
  operator: '=',
  value: 'blank ',
  source: { start: { line: 1, column: 1 }, end: { line: 1, column: 16 } },
  sourceIndex: 0,
  attribute: ' target',
  spaces: { before: '', after: '' },
  type: 'attribute',
  raw: {},
  parent: 
   Selector {
     spaces: { before: '', after: '' },
     nodes: [ [Circular] ],
     type: 'selector',
     parent: 
      Root {
        spaces: [Object],
        nodes: [Object],
        type: 'root',
        lastEach: 1,
        indexes: [Object] },
     lastEach: 1,
     indexes: { '1': 0 } } }
```

Spaces in `attribute` and `value`, it is normal check this in spaces?

Comment (davidtheclark) [inline src/rules/selector-attribute-square-brackets-space-inside/index.js]: Eh, forget it, I guess I was wrong!

File: src/rules/selector-attribute-square-brackets-space-inside/index.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      actual: expectation,
+      possible: [
+        "always",
+        "never",
+      ],
+    })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      selectorParser(selectorTree => {
```
Comment (davidtheclark) [inline src/rules/selector-attribute-square-brackets-space-inside/index.js]: I'd say let's run one quick optimization here: if there is no `[` in the selector, don't bother parsing it. 

Does that seem right?

**Comment by davidtheclark** [comment]
Ok, I had one more comment about a possible optimization. Otherwise LGTM.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+# Head
+
+- Added: `selector-attribute-square-brackets-space-inside` rule.
```
Comment (jeddy3) [inline CHANGELOG.md]: As per https://github.com/stylelint/stylelint/issues/1086#issuecomment-212423114, can we rename this `selector-attribute-brackets-space-inside` please?

File: docs/user-guide/rules.md [hunk 4]
```diff
@@ -130,6 +130,7 @@ Here are all the rules within stylelint, grouped by the [*thing*](http://apps.wo
 
 ### Selector
 
+- [`selector-attribute-square-brackets-space-inside`](../../src/rules/selector-attribute-square-brackets-space-inside/README.md): Require a single space or disallow whitespace on the inside of the square brackets within selectors.
```
Comment (jeddy3) [inline docs/user-guide/rules.md]: "Require a single space or disallow whitespace on the inside of the brackets within attribute selectors."

File: src/rules/selector-attribute-square-brackets-space-inside/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# selector-attribute-square-brackets-space-inside
+
+Require a single space or disallow whitespace on the inside of the square brackets within selectors.
```
Comment (jeddy3) [inline src/rules/selector-attribute-square-brackets-space-inside/README.md]: "Require a single space or disallow whitespace on the inside of the brackets within attribute selectors."

File: src/rules/selector-attribute-square-brackets-space-inside/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+```css
+.foo[ target=_blank ]
```
Comment (jeddy3) [inline src/rules/selector-attribute-square-brackets-space-inside/README.md]:
```css
.foo[ target=_blank ]
/** ↑               ↑
 * The space inside these two brackets */
```

File: src/rules/selector-attribute-square-brackets-space-inside/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ * The space inside these two square brackets */
+```
+
+## Options
+
+`string`: `"always"|"never"`
+
+### `"always"`
+
+There *must always* be a single space inside the parentheses.
```
Comment (jeddy3) [inline src/rules/selector-attribute-square-brackets-space-inside/README.md]: "There *must always* be a single space inside the brackets."

File: src/rules/selector-attribute-square-brackets-space-inside/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+[ target ] {}
+```
+
+```css
+[ target=_blank ] {}
+```
+
+### `"never"`
+
+There *must never* be whitespace on the inside the parentheses.
```
Comment (jeddy3) [inline src/rules/selector-attribute-square-brackets-space-inside/README.md]: `There *must never* be whitespace on the inside the brackets.`

File: src/rules/selector-attribute-square-brackets-space-inside/index.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        selectorTree.eachAttribute(attributeNode => {
+          const attributeSelectorString = attributeNode.toString()
+
+          styleSearch({ source: attributeSelectorString, target: "[" }, match => {
+            const nextCharIsSpace = attributeSelectorString[match.startIndex + 1] === " "
+            if (nextCharIsSpace && expectation === "never") {
+              report({
```
Comment (jeddy3) [inline src/rules/selector-attribute-square-brackets-space-inside/index.js]: Can we minimise the duplication within `report()` by using our idiomatic `complain()` pattern as [seen here](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-whitespace-inside/index.js#L45-L66)?

**Comment by jeddy3** [comment]
Looking very good @evilebottnawi . Thanks, as always, for the comprehensive tests!

I've made a few (minor) requests, and then I think this is ready to go in.

**Comment by evilebottnawi** [comment]
@davidtheclark @jeddy3 ping

**Comment by jeddy3** [comment]
Changes look great!

Feel free to merge after a rebase.

**Pull request merged**